handler/handler.go

```diff
@@ -0,0 +1,330 @@
+		}
+		return requestStatus{
+			Code: http.StatusGatewayTimeout,
+			Msg:  err.Error(),
+			Task: task,
+		}
+	}
+	defer h.parallelUploads.Release()
+
+	logger.Info(
+		"upload starting",
+		slog.Any("task", task),
+	)
+
+	if err := h.uploadFileMultipart(r.Context(), task); err != nil {
+		task.stop()
+		return requestStatus{
+			Code: http.StatusInternalServerError,
+			Msg:  err.Error(),
+			Task: task,
+		}
+	}
+
+	task.stop()
+
+	return requestStatus{
+		Code: http.StatusOK,
+		Msg:  "upload success",
+		Task: task,
+	}
+}
+
+func (h *S3ndHandler) parseRequest(task *uploadTask, r *http.Request) error {
+	{
+		file := r.PostFormValue("file")
+		if file == "" {
+			return fmt.Errorf("missing field: file")
+		}
+
+		if !filepath.IsAbs(file) {
+			return fmt.Errorf("only absolute file paths are supported: %q", html.EscapeString(file))
+		}
+
+		task.File = &file
+	}
+
+	{
+		uriRaw := r.PostFormValue("uri")
+		if uriRaw == "" {
+			return fmt.Errorf("missing field: uri")
+		}
+
+		uri, err := url.Parse(uriRaw)
+		if err != nil {
+			return fmt.Errorf("unable to parse URI: %q", html.EscapeString(uriRaw))
+		}
+
+		if uri.Scheme != "s3" {
+			return fmt.Errorf("only s3 scheme is supported: %q", html.EscapeString(uriRaw))
+		}
+
+		bucket := uri.Host
+		if bucket == "" {
+			return fmt.Errorf("unable to parse bucket from URI: %q", html.EscapeString(uriRaw))
+		}
+
+		key := uri.Path[1:] // Remove leading slash
+
+		task.Uri = &requestURL{*uri}
+		task.Bucket = &bucket
+		task.Key = &key
+	}
+
+	return nil
+}
+
+func (h *S3ndHandler) uploadFileMultipart(ctx context.Context, task *uploadTask) error {
+	file, err := os.Open(*task.File)
+	if err != nil {
+		return errors.Wrapf(err, "Could not open file %v to upload", *task.File)
+	}
+	defer file.Close()
+
+	maxAttempts := *h.conf.UploadTries
+	for task.Attempts = 1; task.Attempts <= maxAttempts; task.Attempts++ {
+		uploadCtx, cancel := context.WithTimeoutCause(ctx, *h.conf.UploadTimeout, errUploadAttemptTimeout)
+		defer cancel()
+		_, err = h.uploader.Upload(uploadCtx, &s3.PutObjectInput{
+			Bucket: aws.String(*task.Bucket),
+			Key:    aws.String(*task.Key),
+			Body:   file,
+		})
+		if err != nil {
+			var apiErr smithy.APIError
+			cause := context.Cause(uploadCtx)
+
+			switch {
+			case errors.As(err, &apiErr):
+				if apiErr.ErrorCode() == "NoSuchBucket" {
+					return errors.Wrapf(err, "upload failed because the bucket %v does not exist", *task.Bucket)
+				}
+			case errors.Is(cause, errUploadAttemptTimeout):
+				errMsg := fmt.Sprintf("upload attempt %v/%v timeout", task.Attempts, maxAttempts)
+
+				// bubble up the error if we've exhausted our attempts
+				if task.Attempts == maxAttempts {
+					return errors.Wrap(err, errMsg)
+				}
+				// otherwise, log the timeout and carry on
+				logger.Warn(errMsg, slog.Any("task", task))
+				continue
+			case errors.Is(err, context.Canceled):
+				// the parent context was cancelled, probalby because the client disconnected
```

Comment:
"probably", and maybe put that into the error message if you're not sure that's why.